### PR TITLE
PGD 5 data node support

### PR DESCRIPTION
**Pull request by vibhorkum**
This PR introduces support for the PGD5 data node with EPAS on Rocky/RHEL/Almalinux. At present, our support is limited to EPAS and PGD5. However, we will add PG Extended in future deployments.

### CODE REVIEW

File: roles/setup_pgd/README.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
+Role Name
```
Comment (mw2q) [inline roles/setup_pgd/README.md:1]: I don't think a README should be added simply as place holder. :)

File: roles/setup_pgd/defaults/main.yml [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+# PGD Cluster Nodes
+pgd_cluster_nodes: []
+
+# PGD DSN variables
+local_node_dsn: "host={{ hostvars[inventory_hostname].private_ip }} port={{ pg_port }} user={{ pgd_replication_user }} dbname={{ pgd_cluster_database }}"
+lead_primary_dsn: ""
+
+etc_hosts_lists: []
+
+supported_os:
```
Comment (mw2q) [inline roles/setup_pgd/defaults/main.yml:116]: The PR message says "support for the PGD5 data node with EPAS on Rocky/RHEL/Almalinux".  I think there are some extra operating systems in here.

File: roles/setup_pgd/defaults/main.yml [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+  - Debian10
+  - Rocky8
+  - AlmaLinux8
+  - OracleLinux7
+
+supported_pg_type:
+  - EPAS
+  - PG
+
+supported_pg_version:
```
Comment (mw2q) [inline roles/setup_pgd/defaults/main.yml:132]: I'm pretty sure v5 doesn't support all of these versions. :)

File: roles/setup_pgd/meta/main.yml [hunk 4]
```diff
@@ -0,0 +1,52 @@
+galaxy_info:
```
Comment (mw2q) [inline roles/setup_pgd/meta/main.yml:1]: Galaxy info should be filled out.

File: roles/setup_pgd/vars/EPAS_Debian.yml [hunk 5]
```diff
@@ -0,0 +1,48 @@
+---
```
Comment (mw2q) [inline roles/setup_pgd/vars/EPAS_Debian.yml:1]: Is the PR message wrong, or do we support debian?

File: roles/setup_pgd/.travis.yml [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+  # Create ansible.cfg with correct roles_path
+  - printf '[defaults]\nroles_path=../' >ansible.cfg
+
+script:
+  # Basic role syntax check
+  - ansible-playbook tests/test.yml -i tests/inventory --syntax-check
+
+notifications:
+  webhooks: https://galaxy.ansible.com/api/v1/notifications/
```
Comment (mw2q) [inline roles/setup_pgd/.travis.yml:29]: Check your editor to make sure the EOF is in there.

File: roles/init_dbserver/tasks/pg_initdb.yml [hunk 7]
```diff
@@ -56,4 +59,4 @@
     daemon_reload: true
     state: started
     enabled: true
-  become: true
+  become: true
```
Comment (mw2q) [inline roles/init_dbserver/tasks/pg_initdb.yml]: Check you editor's EOF setting.  It's producing an unnecessary diff in this instance.

File: roles/manage_dbserver/defaults/main.yml [hunk 8]
```diff
@@ -47,3 +47,6 @@ input_password: ""
 
 pg_databases: []
 params_restart_required: false
+
+# ignore query execution error variable
+ignore_query_execution_error: true
```
Comment (mw2q) [inline roles/manage_dbserver/defaults/main.yml]: Ditto EOF.

File: roles/setup_pgd/defaults/main.yml [hunk 9]
```diff
@@ -134,4 +135,7 @@ supported_pg_version:
   - 12
   - 13
   - 14
-  - 15
\ No newline at end of file
+  - 15
+
+supported_pgd_version:
+  - 5
```
Comment (mw2q) [inline roles/setup_pgd/defaults/main.yml:141]: ditto EOF

File: roles/setup_pgd/tasks/bdr_init_physical.yml [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+  become: true
+
+- name: Update PGD hba config on a new node
+  ansible.builtin.include_role:
+    name: manage_dbserver
+    tasks_from: manage_hba_conf
+  vars:
+    pg_hba_ip_addresses: "{{ pg_allow_ip_addresses }}"
+  no_log: "{{ disable_logging }}"
+  when: not initdb_executed
```
Comment (mw2q) [inline roles/setup_pgd/tasks/bdr_init_physical.yml:55]: ditto EOF

File: roles/setup_pgd/tasks/bdr_init_physical.yml [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
+- name: Create PGD node using bdr_init_physical
+  ansible.builtin.shell: >
+    PGAPPNAME={{ inventory_hostname }} {{ bdr_init_physical }}
+  args:
+    creates: "{{ pg_data }}/PG_VERSION"
+  become: true
+  become_user: "{{ pg_owner }}"
+  async: 180000
```
Comment (mw2q) [inline roles/setup_pgd/tasks/bdr_init_physical.yml:21]: We've been removing these async lines because they are causing issue.  Don't we want to keep removing them?

File: roles/setup_pgd/tasks/pgd_first_node.yml [hunk 12]
```diff
@@ -0,0 +1,8 @@
+---
+- name: Capture PGD lead primary node
+  ansible.builtin.set_fact:
+    lead_primary_node: "{{ lookup('edb_devops.edb_postgres.pgd_lead_primary', hostvars[inventory_hostname].ha_location , wantList=True) }}"
+
+- name: Prepare PGD lead primary DSN
+  ansible.builtin.set_fact:
+    lead_primary_dsn: "host={{ lead_primary_node.private_ip }} port={{ pg_port }} user={{ pgd_replication_user }} dbname={{ pgd_cluster_database }}"
```
Comment (mw2q) [inline roles/setup_pgd/tasks/pgd_first_node.yml:8]: ditto EOF

File: roles/setup_pgd/tasks/pgd_lead_primary_node.yml [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+
+- name: Prepare PGD lead primary DSN using private_ip
+  ansible.builtin.set_fact:
+    lead_primary_dsn: "host={{ lead_primary_node.private_ip }} port={{ pg_port }} user={{ pgd_replication_user }} dbname={{ pgd_cluster_database }}"
+  when: not use_hostname
+
+- name: Prepare PGD lead primary DSN using hostname
+  ansible.builtin.set_fact:
+    lead_primary_dsn: "host={{ lead_primary_node.inventory_hostname }} port={{ pg_port }} user={{ pgd_replication_user }} dbname={{ pgd_cluster_database }}"
+  when: use_hostname
```
Comment (mw2q) [inline roles/setup_pgd/tasks/pgd_lead_primary_node.yml:14]: ditto EOF

File: roles/setup_pgd/tasks/pgd_replication_user.yml [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+- name: Update the pgpass with pgd_replication_user password
+  ansible.builtin.include_role:
+    name: manage_dbserver
+    tasks_from: manage_pgpass
+  vars:
+    pg_pgpass_values:
+      - user: "{{ pgd_replication_user }}"
+        password: "{{ pgd_replication_user_password }}"
+        create: true
+  no_log: "{{ disable_logging }}"
```
Comment (mw2q) [inline roles/setup_pgd/tasks/pgd_replication_user.yml:49]: ditto EOF

File: roles/setup_pgd/tasks/prepare_hba_value_list.yml [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+  run_once: true
+  no_log: "{{ disable_logging }}"
+
+- name: Update PGD node hba config
+  ansible.builtin.include_role:
+    name: manage_dbserver
+    tasks_from: manage_hba_conf
+  vars:
+    pg_hba_ip_addresses: "{{ pg_allow_ip_addresses }}"
+  no_log: "{{ disable_logging }}"
```
Comment (mw2q) [inline roles/setup_pgd/tasks/prepare_hba_value_list.yml:66]: ditto EOF

File: roles/setup_pgd/tasks/update_hba_config.yml [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+  no_log: "{{ disable_logging }}"
+
+- name: Update PGD node hba config
+  ansible.builtin.include_role:
+    name: manage_dbserver
+    tasks_from: manage_hba_conf
+  vars:
+    pg_hba_ip_addresses: "{{ pg_allow_ip_addresses }}"
+  when: initdb_executed
+  no_log: "{{ disable_logging }}"
```
Comment (mw2q) [inline roles/setup_pgd/tasks/update_hba_config.yml:67]: ditto EOF

File: roles/setup_repo/tasks/validate_setup_repo.yml [hunk 17]
```diff
@@ -15,7 +15,7 @@
   register: apt_sources
   changed_when: apt_sources.rc == '0'
 
-- name: Check for EDB Repo on RedHat
+- name: Check for EDB Repo on RedHat 1.0
```
Comment (mw2q) [inline roles/setup_repo/tasks/validate_setup_repo.yml:18]: I find "1.0" and "2.0" confusing.  I think saying "EnterpriseDB" and "enterprisedb-enterprise" is better.

Or is it "EDB v1 repo" and "EDB v2 repo"?

**Review by mw2q — CHANGES_REQUESTED** [review]
I've made a pass and a number of comments.

File: roles/setup_repo/tasks/validate_setup_repo.yml (hunk 17 above)
Comment (vibhorkumar123) [reply roles/setup_repo/tasks/validate_setup_repo.yml:18]: It's EDB v1 repo and EDB v2 repo. Both has different entries for EnterpriseDB Repository.

File: playbook-examples/bdr/inventory-pgd-1.yml [hunk 18]
```diff
@@ -11,15 +11,13 @@ all:
...
-          bdr: []
+          ha_location: PGD_DC1
+          pgd: []
         barman2.home.virt:
           ansible_host: 10.0.0.4
           private_ip: 10.0.0.4
-          etcd: true
-          ha_location: BDRDC2
-          bdr: []
+          ha_location: PGD_DC2
```
[the remark this answers is not in the document]
Comment (vibhorkumar123) [reply playbook-examples/bdr/inventory-pgd-1.yml:19]: I am OK with renaming it. What's your variable name? DC or data_center or location?

File: roles/setup_pgd/.travis.yml [hunk 19]
```diff
@@ -0,0 +1,29 @@
+---
+language: python
+python: "2.7"
```
[the remark this answers is not in the document]
Comment (hannahms) [reply roles/setup_pgd/.travis.yml:3]: We removed all the `.travis.yml` files from the repo in December, I do not think this file should be included

File: roles/setup_pgd/README.md (hunk 1 above)
Comment (vibhorkumar123) [reply roles/setup_pgd/README.md:1]: It's WIP. I haven't finished testing yet. I will add README.

File: roles/setup_pgd/.travis.yml (hunk 19 above)
[the remark this answers is not in the document]
Comment (vibhorkumar123) [reply roles/setup_pgd/.travis.yml:3]: It's for CI.

File: roles/setup_pgd/defaults/main.yml (hunk 2 above)
Comment (vibhorkumar123) [reply roles/setup_pgd/defaults/main.yml:116]: Yes, the goal is to initially begin with the RHEL variant and subsequently expand to other variants

File: roles/setup_repo/tasks/validate_setup_repo.yml (hunk 17 above)
Comment (mw2q) [reply roles/setup_repo/tasks/validate_setup_repo.yml:18]: Ok, then it should probably be EDB Repo 1.0 ...  Because it looks like RedHat 1.0. :)

File: playbook-examples/bdr/inventory-pgd-1.yml (hunk 18 above)
[the remark this answers is not in the document]
Comment (mw2q) [reply playbook-examples/bdr/inventory-pgd-1.yml:19]: i think DC or location are my two favorites.  the two easiest to type.

File: roles/setup_pgd/.travis.yml (hunk 19 above)
[the remark this answers is not in the document]
Comment (vibhorkumar123) [reply roles/setup_pgd/.travis.yml:3]: ok

File: playbook-examples/bdr/inventory-pgd-1.yml (hunk 18 above)
[the remark this answers is not in the document]
Comment (vibhorkumar123) [reply playbook-examples/bdr/inventory-pgd-1.yml]: Changed ha_location => location

File: roles/setup_pgd/.travis.yml (hunk 19 above)
[the remark this answers is not in the document]
Comment (vibhorkumar123) [reply roles/setup_pgd/.travis.yml]: Removed

File: roles/setup_pgd/defaults/main.yml (hunk 2 above)
Comment (vibhorkumar123) [reply roles/setup_pgd/defaults/main.yml:116]: Fixed. Removed the placeholders for other OSs

File: roles/setup_pgd/defaults/main.yml (hunk 3 above)
Comment (vibhorkumar123) [reply roles/setup_pgd/defaults/main.yml:129]: Fixed

File: roles/setup_pgd/tasks/pgd_replication_user.yml (hunk 14 above)
Comment (vibhorkumar123) [reply roles/setup_pgd/tasks/pgd_replication_user.yml:49]: Fixed

File: roles/setup_pgd/tasks/prepare_hba_value_list.yml (hunk 15 above)
Comment (vibhorkumar123) [reply roles/setup_pgd/tasks/prepare_hba_value_list.yml]: Fixed

File: roles/setup_pgd/tasks/update_hba_config.yml (hunk 16 above)
Comment (vibhorkumar123) [reply roles/setup_pgd/tasks/update_hba_config.yml]: FIxed

File: roles/setup_pgd/tasks/verify_pgd_node_group.yml [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+    ignore_query_execution_error: false
+  when: initdb_executed
+
+- name: Verify PGD local node joined group or not
+  ansible.builtin.set_fact:
+     pgd_cluster_group_joined: true
+  when:
+    - sql_query_output.results[0].query_result[0].node_group_name is defined
+    - sql_query_output.results[0].query_result[0].node_group_name == pgd_cluster_name
+    - initdb_executed
```
[the remark this answers is not in the document]
Comment (vibhorkumar123) [reply roles/setup_pgd/tasks/verify_pgd_node_group.yml:38]: FIxed

File: roles/setup_repo/tasks/validate_setup_repo.yml (hunk 17 above)
Comment (vibhorkumar123) [reply roles/setup_repo/tasks/validate_setup_repo.yml]: FIxed

File: plugins/lookup/supported_roles.py [hunk 21]
```diff
@@ -38,6 +38,7 @@
         'manage_dbpatches',
         'manage_efm',
         'setup_patroni',
+        'setup_pgd',
```
Comment (hannahms) [inline plugins/lookup/supported_roles.py:41]: I think the addition of the `setup_pgd` role should not be by default and should mirror the BDR implementation, this can assist with determining when to run the `init_dbserver` role instead of using a flag. Regardless, this file should be cleaned up of BDR items.

File: roles/init_dbserver/defaults/main.yml [hunk 22]
```diff
@@ -5,11 +5,19 @@ pg_version: 14
 pg_ssl: true
 pg_instance_name: main
 pg_remote_ssl_src: false
+
+# This parameter define if we want to run pg_initdb command on a server or not
+run_initdb: true
```
Comment (hannahms) [inline roles/init_dbserver/defaults/main.yml:10]: I think this should be an inventory variable passed through the `pgd` dictionary or pgd role dependent instead of a flag.

File: roles/init_dbserver/tasks/init_dbserver.yml [hunk 23]
```diff
@@ -59,9 +59,17 @@
     - not validate_only|bool
   become: true
 
+- name: Set run_initdb to false if it's a PGD server and has use_physical_backup
+  set_fact:
+    run_initdb: false
+  when:
+    - hostvars[inventory_hostname].pgd.use_physical_backup is defined
+    - hostvars[inventory_hostname].pgd.use_physical_backup|bool
+
```
Comment (hannahms) [inline roles/init_dbserver/tasks/init_dbserver.yml:68]: I think these checks should be within the `supported_roles.py` file

File: roles/init_dbserver/templates/postgresql.conf.template [hunk 24]
```diff
@@ -3,7 +3,7 @@
 unix_socket_directories = '{{ pg_unix_socket_directories | join(",") }}'
 port = {{ pg_port }}
 listen_addresses = '*'
-wal_level = replica
+wal_level = logical
```
Comment (hannahms) [inline roles/init_dbserver/templates/postgresql.conf.template:6]: I think this should be left as `replica` to not break the `setup_replication` role. Another template can be added with `logical` setting, or the `wal_level` can be added as a variable that needs to be validated against `supported_wal_level_values` and templated into this template.

File: roles/setup_pgd/handlers/main.yml [hunk 25]
```diff
@@ -0,0 +1,2 @@
+---
+# handlers file for add_pgdnode
```
Comment (hannahms) [inline roles/setup_pgd/handlers/main.yml:2]: If no handlers are necessary, this directory should be removed.

File: roles/setup_pgd/tasks/pgd_replication_user.yml [hunk 26]
```diff
@@ -0,0 +1,49 @@
...
+  vars:
+    input_user: "{{ pgd_replication_user }}"
+    input_password: "{{ pgd_replication_user_password }}"
+  no_log: "{{ disable_logging }}"
+  when: pgd_replication_user_password|length < 1
+
+- name: Set pgd_replication_user_password
+  ansible.builtin.set_fact:
+    pgd_replication_user_password: "{{ input_password }}"
+  when: pgd_replication_user_password|length < 1
```
Comment (hannahms) [inline roles/setup_pgd/tasks/pgd_replication_user.yml:15]: `no_log: "{{ disable_logging }}"`

File: roles/setup_pgd/tasks/pgd_cluster_database_owner.yml [hunk 27]
```diff
@@ -0,0 +1,29 @@
...
+    input_password: "{{ pgd_cluster_database_owner_password }}"
+  no_log: "{{ disable_logging }}"
+  when:
+    - pgd_cluster_database_owner_password|length < 1
+
+- name: Set pgd_cluster_database_owner_password
+  ansible.builtin.set_fact:
+    pgd_cluster_database_owner_password: "{{ input_password }}"
+  when:
+    - pgd_cluster_database_owner_password|length < 1
```
Comment (hannahms) [inline roles/setup_pgd/tasks/pgd_cluster_database_owner.yml:17]: `no_log: "{{ disable_logging }}"`

File: roles/setup_repo/tasks/setup_repo.yml [hunk 28]
```diff
@@ -32,16 +32,16 @@
...
-    bdr_nodes_ips: "{{ lookup('edb_devops.edb_postgres.bdr_nodes') | flatten | map(attribute='ansible_host') | flatten }}"
+    _ips: "{{ lookup('edb_devops.edb_postgres.pgd_nodes') | flatten | map(attribute='ansible_host') | flatten }}"
 
 - name: Validate tpa_subscription_token
   ansible.builtin.fail:
     msg: "tpa_subscription_token is not defined correctly."
   when:
     - tpa_subscription_token | length < 1
-    - bdr_nodes_ip is defined and bdr_nodes_ips | length > 0
+    - _ip is defined and pgd_nodes_ips | length > 0
```
Comment (hannahms) [inline roles/setup_repo/tasks/setup_repo.yml:44]: where do `_ip` and `pgd_nodes_ips` get defined?

File: plugins/lookup/pgd_lead_primary.py [hunk 29]
```diff
@@ -0,0 +1,94 @@
...
+      - "Retrieves the PGD (Postgres Distributed) lead master for a giving HA location"
+    options:
+      _terms:
+        description: HA location name.
+        required: True
+"""
+
+EXAMPLES = """
+- name: Get the PGD lead primary of the BDRDC1 HA location
+  debug: msg="{{ lookup('[gd_lead_primary', 'BDRDC1') }}"
```
Comment (hannahms) [inline plugins/lookup/pgd_lead_primary.py:20]: `pgd_lead_primary`

File: plugins/lookup/pgd_lead_primary.py [hunk 30]
```diff
@@ -0,0 +1,94 @@
...
+__metaclass__ = type
+
+
+
+DOCUMENTATION = """
+    name: pgd_lead_primary
+    author: Julien Tachoires & Vibhor Kumar
+    short_description: Lookup function for PGD (Postgres Distributed) lead primary node
+    description:
+      - "Retrieves the PGD (Postgres Distributed) lead master for a giving HA location"
```
Comment (hannahms) [inline plugins/lookup/pgd_lead_primary.py:11]: lead primary instead of lead master

File: roles/init_dbserver/templates/postgresql.conf.template (hunk 24 above)
Comment (vibhorkumar123) [reply roles/init_dbserver/templates/postgresql.conf.template:6]: The setting "wal_level = logical" enables streaming replication, logical replication, and PGD. In other words, the "logical" level encompasses all replica usage scenarios, including logical decoding.

File: roles/init_dbserver/tasks/init_dbserver.yml (hunk 23 above)
Comment (vibhorkumar123) [reply roles/init_dbserver/tasks/init_dbserver.yml:68]: Not sure. I am following you. I will ping you to get more information.

File: roles/init_dbserver/defaults/main.yml (hunk 22 above)
Comment (vibhorkumar123) [reply roles/init_dbserver/defaults/main.yml:10]: I thought about that. However, it will create challenges for PSR (Physical Streaming Replication).

Comment (vibhorkumar123) [reply roles/init_dbserver/defaults/main.yml:10]: As discussed. Resolving this.

File: roles/init_dbserver/tasks/init_dbserver.yml (hunk 23 above)
Comment (vibhorkumar123) [reply roles/init_dbserver/tasks/init_dbserver.yml:68]: As discussed, resolving this.

File: roles/init_dbserver/templates/postgresql.conf.template (hunk 24 above)
Comment (vibhorkumar123) [reply roles/init_dbserver/templates/postgresql.conf.template:6]: As discussed. Resolving this conversation.

File: plugins/lookup/pgd_lead_primary.py (hunk 29 above)
Comment (vibhorkumar123) [reply plugins/lookup/pgd_lead_primary.py:20]: Fixed.

File: roles/install_dbserver/defaults/main.yml [hunk 31]
```diff
@@ -3,7 +3,7 @@
 os: ""
 pg_type: "PG"
 pg_version: 14
-bdr_version: 4
+pgd_version: 4
```
Comment (mw2q) [inline roles/install_dbserver/defaults/main.yml]: based on the commit descriptions, shouldn't v5 be the default?

File: roles/setup_pgd/.travis.yml (hunk 6 above)
Comment (mw2q) [reply roles/setup_pgd/.travis.yml]: this file was supposed to be removed based an previous discussion?

File: roles/setup_pgd/vars/edb-ssl.yml [hunk 32]
```diff
@@ -0,0 +1,24 @@
...
+cert_email: "support@enterprisedb.com"
+
+pg_ssl_postgres_conf_params:
+  - {name: "ssl", value: "on"}
+  - {name: "ssl_ca_file", value: "{{ pg_data }}/root.crt"}
+  - {name: "ssl_key_file", value: "{{ pg_data }}/server.key"}
+  - {name: "ssl_cert_file", value: "{{ pg_data }}/server.crt"}
+  - {name: "ssl_crl_file", value: "{{ pg_data }}/root.crl"}
+
+# pg_ssl_ip_addresses:
```
Comment (mw2q) [inline roles/setup_pgd/vars/edb-ssl.yml:16]: I'd prefer commented code to just not be added, unless there is a TODO or FIXME note explaining why it's there.

**Review by mw2q — CHANGES_REQUESTED** [review]

File: roles/install_dbserver/defaults/main.yml (hunk 31 above)
Comment (vibhorkumar123) [reply roles/install_dbserver/defaults/main.yml]: This is already resolved in PR.

File: roles/setup_pgd/vars/edb-ssl.yml (hunk 32 above)
Comment (vibhorkumar123) [reply roles/setup_pgd/vars/edb-ssl.yml:16]: This is intentional. Example of how to use in the edb-ssl.yml

**Review by hannahms — CHANGES_REQUESTED** [review]
-Make sure the EOF blank line exists
-add test case and supporting docker-compose, makefile, test case python script, etc.
-narrow down vars files so each variable is defined either in the main.yml or the OS/pg_type vars file
-narrow down where to store each file, it seems like there are a lot of repeats between setup_pgd and init_dbserver.

File: plugins/lookup/pgd_nodes.py [hunk 33]
```diff
@@ -0,0 +1,88 @@
...
+RETURN = """
+_value:
+  description:
+    - List of PGD nodes
+  type: list
+  elements: dict
+"""
+
+from ansible.plugins.lookup import LookupBase
+from ansible.utils.display import Display
```
[the remark this answers is not in the document]
Comment (vibhorkumar123) [reply plugins/lookup/pgd_nodes.py:35]: We should keep it for all plugins. Display would help in knowing what plugins are doing when user chooses to run -vvvv with ansible command. In future, we should create task to update other plugins.

File: plugins/lookup/supported_roles.py [hunk 34]
```diff
@@ -183,16 +183,17 @@ def run(self, terms, variables=None, **kwargs):
...
-                # (primary, read_only, lead_master, and witness).
-                if len(set(bdr_roles) & set(hostvars['bdr'].get('roles', []))):
+            # PGD cases
+            if (group in ['primary'] and hostvars.get('pgd')):
+                # PGD nodes are a component of the primary group and
+                # must have the 'node_kind' host variable defined.
+                # Additionally, the 'pgd_node_kinds' list should include
+                # at least one of the available PGD node kinds, namely:
+                #data, standby, subscribe-only, and witness.
+                if hostvars['pgd'].get('node_kind', '') in pgd_node_kinds:
```
[the remark this answers is not in the document]
Comment (vibhorkumar123) [reply plugins/lookup/supported_roles.py]: Fixed

File: roles/init_dbserver/tasks/init_dbserver.yml [hunk 35]
```diff
@@ -77,6 +85,7 @@
   when:
     - pg_ssl
     - not validate_only|bool
+    - run_initdb
```
[the remark this answers is not in the document]
Comment (vibhorkumar123) [reply roles/init_dbserver/tasks/init_dbserver.yml:88]: Separate role would be needed to deduplicate code

File: roles/setup_repo/tasks/PG_RedHat_setuprepos.yml [hunk 36]
```diff
@@ -133,14 +133,14 @@
...
-# Adding BDR repository if BDR nodes are defined
+# Adding PGD repository if PGD nodes are defined
 - name: Install curl package if not available
   ansible.builtin.package:
     name: curl
     state: present
   become: true
 
-- name: Install BDR4 packages repo if tpa_subscription_token is given
+- name: Install PGD4 packages repo if tpa_subscription_token is given
```
[the remark this answers is not in the document]
Comment (vibhorkumar123) [reply roles/setup_repo/tasks/PG_RedHat_setuprepos.yml:145]: Separate PR for removing BDR4

File: roles/setup_pgd/tasks/clean_directories.yml [hunk 37]
```diff
@@ -0,0 +1,101 @@
...
+    loop_var: line_item
+  become: true
+  when: pg_unix_socket_directories|length > 0
+
+# TDE specific files clean up
+- name: Remove master key's secure key file if file exists
+  ansible.builtin.file:
+    path: "{{ edb_secure_master_key_file }}"
+    state: absent
+  become: true
```
[the remark this answers is not in the document]
Comment (vibhorkumar123) [reply roles/setup_pgd/tasks/clean_directories.yml:89]: Fixed

File: roles/setup_pgd/tasks/clean_directories.yml [hunk 38]
```diff
@@ -0,0 +1,101 @@
...
+
+- name: Remove master key file if file exists
+  ansible.builtin.file:
+    path: "{{ edb_master_key_plain_file }}"
+    state: absent
+  become: true
```
[the remark this answers is not in the document]
Comment (vibhorkumar123) [reply roles/setup_pgd/tasks/clean_directories.yml:96]: Fixed

File: roles/setup_pgd/tasks/clean_directories.yml [hunk 39]
```diff
@@ -0,0 +1,101 @@
...
+
+- name: Remove master key directory if directory exists
+  ansible.builtin.file:
+    path: "{{ edb_master_key_location }}"
+    state: absent
+  become: true
```
[the remark this answers is not in the document]
Comment (vibhorkumar123) [reply roles/setup_pgd/tasks/clean_directories.yml:103]: fixed

File: roles/setup_pgd/tasks/pgd_lead_primary_node.yml [hunk 40]
```diff
@@ -0,0 +1,40 @@
+---
+- name: Capture PGD lead primary node
+  ansible.builtin.set_fact:
+    lead_primary_node: "{{ lookup('edb_devops.edb_postgres.pgd_lead_primary', hostvars[inventory_hostname].location , wantList=True) }}"
+
+- name: Prepare PGD lead primary DSN using private_ip
+  ansible.builtin.set_fact:
+    lead_primary_dsn: "host={{ lead_primary_node.private_ip }} port={{ pg_port }} user={{ pgd_replication_user }} dbname={{ pgd_cluster_database }}"
```
[the remark this answers is not in the document]
Comment (vibhorkumar123) [reply roles/setup_pgd/tasks/pgd_lead_primary_node.yml:8]: Not an issue.

File: playbook-examples/pgd/playbook-bdr-lookup-functions.yml
Comment (hannahms) [inline playbook-examples/pgd/playbook-bdr-lookup-functions.yml:1]: I don't think this is necessary to include in the repo

**Review by hannahms — CHANGES_REQUESTED** [review]
Just a few minor changes noted in comments

File: tests/cases/setup_pgd/inventory.yml
[the remark this answers is not in the document]
Comment (vibhorkumar123) [reply tests/cases/setup_pgd/inventory.yml:1]: Yeah. Remnant from testing

File: tests/cases/setup_pgd/Makefile
[the remark this answers is not in the document]
Comment (vibhorkumar123) [reply tests/cases/setup_pgd/Makefile:1]: I thought I had fixed it. Let me re-check

File: roles/setup_pgd/tasks/validate_setup_pgd.yml [hunk 41]
```diff
@@ -0,0 +1,195 @@
...
+  ansible.builtin.service_facts:
+
+- name: Check if service pg_service is running
+  ansible.builtin.assert:
+    that:
+      - ansible_facts.services[pg_service + '.service']['state'] == 'running'
+    fail_msg: "The service {{ pg_service }} is not running."
+    success_msg: "The service {{ pg_service }} is running."
+
+- name: Check if service pg_service is running
```
[the remark this answers is not in the document]
Comment (vibhorkumar123) [reply roles/setup_pgd/tasks/validate_setup_pgd.yml:52]: Copy from `init_dbserver/tasks/validate_*`

File: playbook-examples/pgd/playbook-bdr-lookup-functions.yml
Comment (vibhorkumar123) [reply playbook-examples/pgd/playbook-bdr-lookup-functions.yml:1]: ok

File: roles/setup_pgd/tasks/validate_setup_pgd.yml (hunk 41 above)
[the remark this answers is not in the document]
Comment (vibhorkumar123) [reply roles/setup_pgd/tasks/validate_setup_pgd.yml:52]: Why do you wish to modify the task name? It appears suitable as it is.

File: tests/cases/setup_pgd/Makefile
[the remark this answers is not in the document]
Comment (vibhorkumar123) [reply tests/cases/setup_pgd/Makefile:1]: Fixed

File: roles/setup_pgd/tasks/validate_setup_pgd.yml (hunk 41 above)
[the remark this answers is not in the document]
Comment (hannahms) [reply roles/setup_pgd/tasks/validate_setup_pgd.yml:52]: It is checking if it is enabled, the task prior has the same name and checks if it is running

**Review by hannahms — APPROVED** [review]
LGTM
